test/extended/openstack/utils.go

```diff
@@ -108,6 +141,147 @@ func DeleteNamespace(ctx context.Context, clientSet *kubernetes.Clientset, ns *v
 	}
 }
 
+func CreatePVC(ctx context.Context, clientSet *kubernetes.Clientset, baseName string, nsName string, scName string, pvcSize string) *v1.PersistentVolumeClaim {
+	pvcName := fmt.Sprintf("%v-%v", baseName, RandomSuffix())
+	storageClassName := scName
+	pvc := &v1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      pvcName,
+			Namespace: nsName,
+		},
+		Spec: v1.PersistentVolumeClaimSpec{
+			AccessModes:      []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce},
+			StorageClassName: &storageClassName,
+			Resources: v1.ResourceRequirements{
+				Requests: v1.ResourceList{
+					v1.ResourceStorage: resource.MustParse(pvcSize),
+				},
+			},
+		},
+	}
+
+	_, err := clientSet.CoreV1().PersistentVolumeClaims(nsName).Create(ctx, pvc, metav1.CreateOptions{})
+
+	if err != nil {
+		e2e.Failf("unable to create PersistentVolumeClaim %v: %v", pvc.Name, err)
+		return nil
+	}
+	e2e.Logf("PersistentVolumeClaim %v was created", pvc.ObjectMeta.Name)
+
+	return pvc
+}
+
+func waitPvcVolume(ctx context.Context, clientSet *kubernetes.Clientset, pvc string, ns string) error {
+	err := wait.PollUntilContextTimeout(ctx, 15*time.Second, 3*time.Minute, true, func(ctx context.Context) (bool, error) {
+		pvc, err := clientSet.CoreV1().PersistentVolumeClaims(ns).Get(ctx, pvc, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		} else {
+			if pvc.Spec.VolumeName != "" {
+				return true, nil
+			} else {
+				return false, nil
+			}
+		}
+	})
+	return err
+}
+
+// return share from openstack with specific name
+func GetSharesFromName(client *gophercloud.ServiceClient, shareName string) ([]shares.Share, error) {
+	var emptyShare []shares.Share
+
+	listOpts := shares.ListOpts{
+		Name: shareName,
+	}
+	allPages, err := shares.ListDetail(client, listOpts).AllPages()
+	if err != nil {
+		return emptyShare, err
+	}
+	shares, err := shares.ExtractShares(allPages)
+	if err != nil {
+		return emptyShare, err
+	}
+	return shares, nil
+}
+
+func FindStorageClassByProvider(oc *exutil.CLI, provisioner string) *storagev1.StorageClass {
+	scList, err := oc.AdminKubeClient().StorageV1().StorageClasses().List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			e2e.Logf("no storage classes found")
+			return nil
+		}
+		e2e.Failf("could not list storage classes: %v", err)
+	}
+	for _, sc := range scList.Items {
+		//if strings.HasPrefix(sc.Provisioner, provisionerPrefix) {
+		//		e2e.Logf("ignoring in-tree storage class %s", sc.Name)
+		//		continue
+		//}
+		if sc.Provisioner == provisioner {
+			return &sc
+		}
+	}
+	e2e.Logf("no default storage class found")
+	return nil
+}
+
+// Creates *appsv1.Deployment using the image agnhost configuring a server on the specified port and protocol
+// Further info on: https://github.com/kubernetes/kubernetes/blob/master/test/images/agnhost/README.md#netexec
+func createTestDeployment(deploymentOpts *DeploymentOpts) *appsv1.Deployment {
```

Comment:
If I understand correctly, `deploymentOpts` is mandatory here: you can drop the pointer to avoid nil pointer exceptions.

Also since `createTestDeployment` is not exported, its options should be private too.

```suggestion
func createTestDeployment(deploymentOpts deploymentOpts) *appsv1.Deployment {
```

Reply:
Done